unicore/ema.py

```diff
@@ -40,13 +40,22 @@ def flatten_parameters(self, model, init_param):
         if init_param is not None:
             assert torch.allclose(init_param, flatten_param), "ema init error!"
         torch.cuda.empty_cache()
-        return flatten_param
+        return name2param, flatten_param
 
-    def update(self, new_param):
-        with torch.no_grad():
-            diff = self.param - new_param
-            diff *= 1 - self.decay
-            self.param -= diff
+    def update_one_param(self, ema_param, new_param):
+        diff = ema_param - new_param
+        diff *= 1 - self.decay
+        ema_param -= diff
+
+    def update(self, new_param, is_flattened):
+        if is_flattened:
+            with torch.no_grad():
+                self.update_one_param(self.param, new_param)
+        else:
+            with torch.no_grad():
+                for n, p in new_param:
+                    if n in self.name2param:
```

Comment:
what if n not in "n in self.name2param:" ?

Reply:
I guess parameters not in `self.name2param` are those with requires_grad=False

Reply:
I can also implement this logic in the trainer, by filtering the `model.named_parameters()`. Which implementation is better?